### DISCUSSION TITLE
"Task timed out" error on some Laravel endpoints

### DISCUSSION THREAD

**Discussion by nickrupert7** [brefphp/bref#1002] [Support] [answered]
**Description:**

We have recently migrated one of our Laravel 8 projects to use serverless deployment with Bref on AWS Lambda.

SOME of our endpoints are consistently failing with the error `{"message":"Internal Server Error"}` (`Task timed out after 28.03 seconds` in CloudWatch logs) while other endpoints work perfectly. 

I have seen https://github.com/brefphp/bref/issues/475 but I am not convinced that the error has to do with the database being inaccessible due to our VPC, because again, other endpoints which use the database seem to have no such connection issues. It is also worth noting that our `serverless.yml` file _is_ configured to host the lambdas in the VPC. 

I'm also not convinced that the error is a failure to connect to some other external resource, because logging indicates that the code is crashing before `boot` in `AppServiceProvider` is called, which means that the error is not caused by any code specifically within the failing endpoints. 

The nature of this error suggests that it should be happening on _every_ request, or not at all, which is why it's weird to me that it's only happening on SOME endpoints. And again, it's consistent about which endpoints work and don't work. I also haven't noticed any sort of pattern that links the consistently failing endpoints together.

**How to reproduce:**

_I'm not sure what is causing the issue, so I can't provide generic steps to reproduce._

**Comment by mnapoli** [comment, marked as answer]
Hey! Just to be sure, are you aware of this limitation: https://bref.sh/docs/environment/database.html#accessing-the-internet

**Reply by nickrupert7** [reply]
I am aware of the internet access limitation, but like I indicated in the initial report, it's weird to me that this is only failing on SOME endpoints. Plus it's failing before my endpoint code even gets called - regardless of whether my endpoints try to access the internet or not.

**Reply by nickrupert7** [reply]
I am working on doing additional logging on my endpoints, and trying to determine _for sure_ whether the endpoints that are failing are trying to access the internet. I'm just perplexed by the fact that the app doesn't even appear to load.

Perhaps there's a service registered with Laravel that's trying to access the internet? However, if that's the case, why wouldn't it fail _every time_?

Thank you for responding so quickly, and for working with me to solve this issue!

**Reply by t-richard** [reply]
> Perhaps there's a service registered with Laravel that's trying to access the internet? However, if that's the case, why wouldn't it fail every time?

If this service accesses the internet in its contructor or in a bootstrapping method, it might cause this kind of issues. Because laravel will most likely not instanciate all services in your app but only the one required to fulfill the request. 

Also I had a problem once where the Lambda security group only allowed 443 port (https) but for some weird reason needed 80 (http) in some cases, like all called URLs were HTTPS but it still required port 80 to be allowed :upside_down_face: . 

Maybe try to allow all outbound on the security group if it's not already the case to see if it changes something.

**Reply by nickrupert7** [reply]
I've done some more investigating, and adding a `dd('CHECKPOINT')` statement at the beginning of my endpoint DOES cause it to produce the expected result.

What's weird is that I had a `Log::info('CHECKPOINT')` in my endpoint, and that log message doesn't make it to CloudWatch. I put the same sort of logging on other endpoints that I know to work, and they _did_ show up in CloudWatch, which confirms that I do have Logging configured correctly. I wonder if logs don't get pushed unless the request actually completes?

All of that to say, the fact that my attempts at logging weren't showing up in CloudWatch led me to believe the endpoints weren't getting activated at all, but new evidence suggests otherwise, and the missing log messages were actually a quirk of how Logging works.

**Reply by nickrupert7** [reply]
> Also I had a problem once where the Lambda security group only allowed 443 port (https) but for some weird reason needed 80 (http) in some cases, like all called URLs were HTTPS but it still required port 80 to be allowed 🙃 .

@t-richard That sounds frustrating, but makes a lot of sense, think! I'll definitely try that if i'm still having issues after i set up the NAT gateway.

**Reply by mnapoli** [reply]
I often see that problem ("random" timeouts) with Sentry (or anything similar). On error, Sentry tries to report the error to the online service and since there's no NAT gateway you get a timeout. And it only happens on errors, so it seems random. Just an idea.

**Reply by mnapoli** [reply]
What I usually recommend too is to configure _all_ timeouts to something like 5 seconds. Most libraries have an infinite timeout, which is very bad unfortunately. Most of them can be configured, but unfortunately not all of them.

**Reply by t-richard** [reply]
Yes, there is no real way to get around it all. Sometimes, setting `default_socket_timeout` in INI file can help, also setting cURL timeout option globally can get around libraries not allowing to set the timeout directly.

**Reply by nickrupert7** [reply]
Adding the NAT gateway fixed my issues! I didn't even have to mess with the security groups, or timeout settings.

Would it be possible to publish this conversation publicly so somebody else in the future might find it and benefit from it?

The logging situation is still super weird to me, and the fact that log messages on failing endpoints weren't going to CloudWatch at all led me to incorrect conclusions about what was happening, so again, if somebody else comes along in the future with a similar issue I hope that this conversation might help them in some small way.

In conclusion, I think that the log messages must go to some kind of "buffer" in the Lambda function memory, and that buffer doesn't get dumped to CloudWatch until the request exists - even if the response is an error. But because it's timing out, the function itself never actually issues a response, and so the log buffer never gets dumped. At least I THINK that's what's happening.

**Reply by nickrupert7** [reply]
Also, I tried to mark your answer as the accepted answer, but GitHub isn't letting me. The button doesn't even seem to do anything. 🤷‍♂️

**Reply by mnapoli** [reply]
> Adding the NAT gateway fixed my issues!

> Would it be possible to publish this conversation publicly so somebody else in the future might find it and benefit from it?

😅 This is why I added a whole section in the documentation about that, with a big "WARNING" at the top:

https://bref.sh/docs/environment/database.html#accessing-the-internet

**Reply by nickrupert7** [reply]
Yeah that's my bad for not reading it! Like i said, the thing that really threw me off was that my logging wasn't working, so I really was convinced that the endpoints weren't even running at all. With trouble "accessing the internet" I wouldn't expect the code to hang up until the line that _actually_ calls the internet. 

But alas, you were right, and I thank you for your patience!
